include/net/mqtt.h

```diff
@@ -1,474 +1,685 @@
-	 * will continue.
-	 * Any other value will stop the QoS handshake and the caller will
-	 * return -EINVAL. The application must discard all the messages
-	 * already processed.
-	 *
-	 * <b>Note: this callback must be not NULL</b>
-	 *
-	 * @param [in] ctx MQTT context
-	 * @param [in] pkt_id Packet Identifier for the input MQTT msg
-	 * @param [in] type Packet type
+	MQTT_UNACCEPTABLE_PROTOCOL_VERSION      = 0x01,
+
+	/** The Client identifier is correct UTF-8 but not allowed by the
+	 *  Server.
 	 */
-	int (*publish_tx)(struct mqtt_ctx *ctx, u16_t pkt_id,
-			  enum mqtt_packet type);
-
-	/** Callback executed when a MQTT_APP_SUBSCRIBER,
-	 * MQTT_APP_PUBLISHER_SUBSCRIBER or MQTT_APP_SERVER applications receive
-	 * a MQTT PUBxxxx msg. If this callback returns 0, the caller will
-	 * continue. If type is MQTT_PUBREL this callback must return 0 if
-	 * pkt_id matches the packet id of a previously received MQTT_PUBxxx
-	 * message. Any other value will stop the QoS handshake and the caller
-	 * will return -EINVAL
-	 *
-	 * <b>Note: this callback must be not NULL</b>
-	 *
-	 * @param [in] ctx MQTT context
-	 * @param [in] msg Publish message, this parameter is only used
-	 *                 when the type is MQTT_PUBLISH
-	 * @param [in] pkt_id Packet Identifier for the input msg
-	 * @param [in] type Packet type
+	MQTT_IDENTIFIER_REJECTED                = 0x02,
+
+	/** The Network Connection has been made but the MQTT service is
+	 *  unavailable.
 	 */
-	int (*publish_rx)(struct mqtt_ctx *ctx, struct mqtt_publish_msg *msg,
-			  u16_t pkt_id, enum mqtt_packet type);
-
-	/** Callback executed when a MQTT_APP_SUBSCRIBER or
-	 * MQTT_APP_PUBLISHER_SUBSCRIBER receives the MQTT SUBACK message
-	 * If this callback returns 0, the caller will continue. Any other
-	 * value will make the caller return -EINVAL.
-	 *
-	 * <b>Note: this callback must be not NULL</b>
-	 *
-	 * @param [in] ctx MQTT context
-	 * @param [in] pkt_id Packet Identifier for the MQTT SUBACK msg
-	 * @param [in] items Number of elements in the qos array
-	 * @param [in] qos Array of QoS values
+	MQTT_SERVER_UNAVAILABLE                 = 0x03,
+
+	/** The data in the user name or password is malformed. */
+	MQTT_BAD_USER_NAME_OR_PASSWORD          = 0x04,
+
+	/** The Client is not authorized to connect. */
+	MQTT_NOT_AUTHORIZED                     = 0x05
+};
+
+/** @brief MQTT SUBACK return codes. */
+enum mqtt_suback_return_code {
+	/** Subscription with QoS 0 succeeded. */
+	MQTT_SUBACK_SUCCESS_QoS_0 = 0x00,
+
+	/** Subscription with QoS 1 succeeded. */
+	MQTT_SUBACK_SUCCESS_QoS_1 = 0x01,
+
+	/** Subscription with QoS 2 succeeded. */
+	MQTT_SUBACK_SUCCESS_QoS_2 = 0x02,
+
+	/** Subscription for a topic failed. */
+	MQTT_SUBACK_FAILURE = 0x80
+};
+
+/** @brief Abstracts UTF-8 encoded strings. */
+struct mqtt_utf8 {
+	u8_t *utf8;             /**< Pointer to UTF-8 string. */
+	u32_t size;             /**< Size of UTF string, in bytes. */
```

Comment:
I still find it unfortunate that "mqtt_utf8" has "size", while "mqtt_binstr" - "len".

(Not requesting to fix anything.)

Reply:
For example, all of struct mqtt_puback_param, struct mqtt_pubrec_param, struct mqtt_pubrel_param, struct mqtt_pubcomp_param managed to end up with the member named message_id, but for some reason, mqtt_utf8 and mqtt_binstr have got to have it different.